### PR TITLE
feat(ctl): Add support for `stacklet credentials` subcommand

### LINKED ISSUE #107: Add credential output to `stackablectl`
Add credential output to `stackablectl` based on these requirements:

- Opt-in, no display by default. Introduce a flag like `--show-credentials`
- Handle error cases when the user has no rights to view secrets
- Handle error cases when there are no such secrets, for example when using OAuth/OIDC

### PR DESCRIPTION
**Pull request by Techassi**
Tracked by #98, closes #107 

This adds support for credential output when using. They are only displayed when the user explicitly want to retrieve them via a dedicated subcommand. Using a subcommand makes this an opt-in, as the output contains sensitive data. Test it with:

```shell
cargo run -p stackablectl -- stacklets credentials superset
```

The implementation is not final and will change with the introduction of Discovery 2.0.

---

~Currently blocked by #106~

~Currently blocked by #113~

### CODE REVIEW

**Comment by Techassi** [comment]
To conditionally display the `CREDENTIALS` column, https://github.com/Nukesor/comfy-table/pull/126 is required. As a workaround, we could also duplicate the table header code, once with and once without the credentials column.

**Review by nightkr — COMMENTED** [review]
On second thought, I'm not sure `list` is the right place to display this.. I think it might be more appropriate to have a "show creds for this specific stacklet" command?

File: rust/stackable-cockpit/src/platform/stacklet/mod.rs [hunk 1]
```diff
@@ -112,15 +116,25 @@ async fn list_stackable_stacklets(
...
 
-            let endpoints = get_service_endpoints(product_name, &object_name, &object_namespace)
-                .await
-                .context(ServiceSnafu)?;
+            let endpoints =
+                get_service_endpoints(kube_client, product_name, &object_name, &object_namespace)
+                    .await
+                    .context(ServiceSnafu)?;
+
+            let credentials = match get_credentials(kube_client, product_name, &object).await {
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/platform/stacklet/mod.rs:124]: Yes I thought about this, but didn't implement it in the first iteration of this feature. I will add this asap.

File: rust/stackable-cockpit/src/platform/credentials.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+/// Retrieves the credentials looking up a secret identified by `secret_name`
+/// in `secret_namespace`. The function returns [`Ok(None)`] if `username_key`
+/// and/or `password_key` are not found or the product does not provide
+/// any credentials.
+pub async fn get_credentials(
+    kube_client: &KubeClient,
+    product_name: &str,
+    product_crd: &DynamicObject,
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/platform/credentials.rs]: Changed in cf22bc0

File: rust/stackable-cockpit/src/platform/credentials.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    // FIXME (Techassi): Add a separate CredentialsError which indicates what
+    // went wrong when retrieving credentials:
+    // - No permissions to do so, most likely a 403
+    // - Secret not found, most likely a 404
+    // - No credentialsSecret present, None below
+    // - Username and/or password key not found, don't return Option
+
+    let credentials = match product_name {
+        "airflow" | "superset" => {
+            match product_crd.data["spec"]["clusterConfig"]["credentialsSecret"].as_str() {
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/platform/credentials.rs]: It is...

[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/platform/credentials.rs]: Updated in cf22bc0

File: rust/stackable-cockpit/src/platform/credentials.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                Some(secret_name) => kube_client
+                    .get_credentials_from_secret(
+                        secret_name,
+                        &product_crd.namespace().unwrap(),
+                        "adminUser.username",
+                        "adminUser.password",
+                    )
+                    .await
+                    .context(KubeSnafu)?,
+                None => return Err(NoSecretSnafu.build()),
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/platform/credentials.rs]: Changed in cf22bc0

File: rust/stackable-cockpit/src/utils/k8s/client.rs [hunk 5]
```diff
@@ -168,31 +174,28 @@ impl KubeClient {
...
         username_key: &str,
-        password_key: Option<&str>,
-    ) -> Result<Option<(String, String)>> {
+        password_key: &str,
+    ) -> Result<Credentials> {
         let secret_api: Api<Secret> = Api::namespaced(self.client.clone(), secret_namespace);
 
         let secret = secret_api.get(secret_name).await.context(KubeSnafu)?;
-        let secret_data = secret.data.ok_or(InvalidSecretDataSnafu {}.build())?;
+        let secret_data = secret.data.ok_or(InvalidSecretDataSnafu.build())?;
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/utils/k8s/client.rs]: Changed in cf22bc0

File: rust/stackable-cockpit/src/utils/k8s/client.rs [hunk 6]
```diff
@@ -168,31 +174,28 @@ impl KubeClient {
...
 
         let username = match secret_data.get(username_key) {
             Some(username) => {
                 String::from_utf8(username.0.clone()).context(ByteStringConvertSnafu)?
             }
-            None => return Ok(None),
+            None => return Err(NoUsernameKeySnafu.build()),
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/utils/k8s/client.rs]: Changed in cf22bc0

File: rust/stackable-cockpit/src/utils/k8s/client.rs [hunk 7]
```diff
@@ -168,31 +174,28 @@ impl KubeClient {
...
 
-        let password = match password_key {
-            Some(key) => match secret_data.get(key) {
-                Some(password) => {
-                    String::from_utf8(password.0.clone()).context(ByteStringConvertSnafu)?
-                }
-                None => return Ok(None),
-            },
-            None => REDACTED_PASSWORD.to_string(),
+        let password = match secret_data.get(password_key) {
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/utils/k8s/client.rs]: Changed in cf22bc0

**Comment by sbernauer** [comment]
For the record: https://github.com/stackabletech/stackable-cockpit/commit/a52b933aa51031baaa5b4760cdf0b83c72fd9d9a fixes the 404 error

File: rust/stackable-cockpit/src/platform/stacklet/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/platform/stacklet/mod.rs]: Credential output is now handled by a separate `stacklet` subcommand. The secret is only fetched when executing this command.

File: rust/stackable-cockpit/src/platform/stacklet/mod.rs [hunk 8]
```diff
@@ -41,11 +46,14 @@ pub struct Stacklet {
...
     pub conditions: Vec<DisplayCondition>,
+
+    /// Optional credentials consisting of username and password.
+    pub credentials: Option<Credentials>,
 }
 
 #[derive(Debug, Snafu)]
 pub enum StackletError {
-    #[snafu(display("kubernetes error"))]
+    #[snafu(display("kubernetes error"), context(false))]
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/platform/stacklet/mod.rs:51]: Generally agreed, but maybe let's tackle this in #55

File: rust/stackable-cockpit/src/platform/stacklet/mod.rs [hunk 9]
```diff
@@ -112,15 +116,25 @@ async fn list_stackable_stacklets(
...
+            let endpoints =
+                get_service_endpoints(kube_client, product_name, &object_name, &object_namespace)
+                    .await
+                    .context(ServiceSnafu)?;
+
+            let credentials = match get_credentials(kube_client, product_name, &object).await {
+                Ok(credentials) => credentials,
+                Err(err) => match err {
+                    CredentialsError::KubeError { source } => return Err(source.into()),
+                    CredentialsError::NoSecret => None,
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/platform/stacklet/mod.rs]: I feel like the `get_credentials` function should report this error. The implementer (stackablectl) chooses to map it to `None`, which indicates that no credentials output should be printed.

File: rust/stackable-cockpit/src/platform/stacklet/mod.rs [hunk 10]
```diff
@@ -112,15 +116,25 @@ async fn list_stackable_stacklets(
...
+                },
+            };
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackable-cockpit/src/platform/stacklet/mod.rs]: Changed in 6afb3d8

File: rust/stackablectl/src/cli/mod.rs [hunk 11]
```diff
@@ -210,7 +210,7 @@ pub enum Commands {
 
     /// Interact with deployed stacklets, which are bundles of resources and
     /// containers required to run the product.
-    #[command(aliases(["stacklet", "stl", "sl"]))]
+    #[command(aliases(["stl", "sl"]))]
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackablectl/src/cli/mod.rs:213]: Usually CLI tools use the singular form for resources, like `docker image ls` or `nmcli device status`. That's why I opted to stick with `stacklet`.

We could however introduce a new subcommand `stacklets`, which is basically an alias for `stacklet ls` (like `docker images`).

File: rust/stackablectl/src/cmds/stacklet.rs [hunk 12]
```diff
@@ -16,26 +20,56 @@ use crate::{
...
     /// type 'plain'.
     #[arg(short = 'c', long = "color")]
     use_color: bool,
 
-    #[arg(short, long = "output", value_enum, default_value_t = Default::default())]
+    /// Display credentials for various endpoints. This requires permissions to
+    /// read Kubernetes secrets. These credentials provide access to deployed
+    /// stacklets and thus should be handled with care.
+    #[arg(long)]
+    show_credentials: bool,
```
[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackablectl/src/cmds/stacklet.rs:72]: Yes, this can be removed. Slipped through the crack.

[the remark this answers is not in the document]
Comment (Techassi) [reply rust/stackablectl/src/cmds/stacklet.rs]: Removed in 5e9a77f

File: rust/stackablectl/src/cli/mod.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (sbernauer) [reply rust/stackablectl/src/cli/mod.rs:213]: That actually sounds like a good idea :)
Burt maybe later at some point of time

**Review by sbernauer — APPROVED** [review]
Command worked for me :+1:
```
➜  stackable-cockpit git:(feat/credential-output) cargo r -p stackablectl -- stacklet credentials --product-namespace sbernauer superset simple-superset
Credentials for superset (simple-superset) in namespace 'sbernauer':

 USERNAME  admin 
 PASSWORD  admin
```
In case i changed `stacklet.data["spec"]["clusterConfig"]["credentialsSecret"]` to `stacklet.data["spec"]["clusterConfig"]["credentialsSecret123"]` it showed `No credentials` and did not give an error.
This is good, as we will remove that field in the future.

**Pull request merged**